### DISCUSSION TITLE
How to change Physarum background color?

### DISCUSSION THREAD

**Discussion by Vamoss** [amandaghassaei/glcompute#7] [Q&A]
Hello again Amanda,

I am trying to change the default black background color without success.

We can also refer to this example in the further code:
https://openprocessing.org/sketch/1760726

I tried to change the render fragment shader from this:
```glsl
in vec2 v_uv;
uniform sampler2D u_trail;
uniform float u_scale;
out vec4 out_state;
void main() {
	out_state = vec4(texture(u_trail, v_uv).xyz * u_scale, 1);
}
```

To this:
```glsl
in vec2 v_uv;
uniform sampler2D u_trail;
uniform float u_scale;
out vec4 out_state;
float luma(vec3 color) {
	return dot(color, vec3(0.299, 0.587, 0.114));
}
void main() {
	vec3 trailColor = texture(u_trail, v_uv).xyz * u_scale;
	float brightness = luma(trailColor );
	vec3 backgroundColor = vec3(1.0, 0.0, 0.0);
	vec3 mixedColor = mix(backgroundColor, trailColor , brightness);
	out_state = vec4(vec3(mixedColor), 1);
}
```

The result is almost there, but the mixedColor borders are not smooth.

Thank you once again,
Carlos

**Comment by amandaghassaei** [comment]
That seems right, can you describe the unintended behavior more?  This is what I'm seeing:
<img width="1395" alt="Screen Shot 2023-01-27 at 11 29 49 AM" src="https://user-images.githubusercontent.com/3218981/215180468-7b82fea2-f2f3-4871-980d-9ee1362b1b15.png">

**Comment by Vamoss** [comment]
Thank you for taking a look.

This red over red shouldn't have this black gradient at the borders:
![image](https://user-images.githubusercontent.com/245841/215192803-db532aeb-771d-4e68-b678-bb53374b4bf5.png)

I understand that from the trail data, black means no "food" for the physarum.
If it requires too many changes in the code, no problem, I can live with the gradient :)

**Comment by amandaghassaei** [comment]
I think this effect makes sense given the code bc the trailColor is darker at the edges, so the mix of the dark red with the red gives you a slightly dark red.  Are you thinking that the result should be flat red instead?  For that you could try alpha compositing (using the luma value as alpha) rather than the mix() function: https://en.wikipedia.org/wiki/Alpha_compositing
